### PR TITLE
duplicate check for github usernames in registry.yaml

### LINKED ISSUE #108: The CI check doesn't run if the user has already a contribution the repo
The checks should have run and failed in this PR #107

### PR DESCRIPTION
**Pull request by haliliceylan**
fix #108

### CODE REVIEW

File: ci/check-registry.js [hunk 1]
```diff
@@ -18,10 +18,21 @@ async function validateRegistry() {
...
         throw new Error('Invalid registry format. The "users" key is missing.');
     }
+
+    // Verify that each github username has only one prefix/domain 
+    const ghUsernames = Object.values(registry.users).map(user => user.github_username);
+    const duplicateGhUsernames = ghUsernames.filter((ghUsername, index) => ghUsernames.indexOf(ghUsername) != index)
+
+    if (duplicateGhUsernames.length > 0) {
+        console.log(duplicateGhUsernames)
+        throw new Error('There are duplicate github usernames in the array');
```
Comment (esau-morais) [inline ci/check-registry.js:33]: Typo: `duplicate` must be in the past tense

File: ci/check-registry.js [hunk 2]
```diff
@@ -18,10 +18,21 @@ async function validateRegistry() {
...
     const registry = yaml.parse(registryFileContent);
+
     // Validate the structure of the registry
     if (!registry || !registry.users) {
         throw new Error('Invalid registry format. The "users" key is missing.');
     }
+
+    // Verify that each github username has only one prefix/domain 
+    const ghUsernames = Object.values(registry.users).map(user => user.github_username);
+    const duplicateGhUsernames = ghUsernames.filter((ghUsername, index) => ghUsernames.indexOf(ghUsername) != index)
```
Comment (esau-morais) [inline ci/check-registry.js:29]: Could change to `duplicatedGHUsernames`

**Comment by fatih-yavuz** [comment]
@haliliceylan thank you for the contribution. Do you mind putting this check into a function and adding a test for it? If you need help, I and (I am pretty sure) @bsoyka can help you

**Review by esau-morais — APPROVED** [review]

**Comment by esau-morais** [comment]
> @haliliceylan thank you for the contribution. Do you mind putting this check into a function and adding a test for it? If you need help, I and (I am pretty sure) @bsoyka can help you

I enforce Fatih's suggestion

**Comment by fatih-yavuz** [comment]
> > @haliliceylan thank you for the contribution. Do you mind putting this check into a function and adding a test for it? If you need help, I and (I am pretty sure) @bsoyka can help you
> 
> I enforce Fatih's suggestion

@esau-morais what do you mean by that?

**Comment by esau-morais** [comment]
> > > @haliliceylan thank you for the contribution. Do you mind putting this check into a function and adding a test for it? If you need help, I and (I am pretty sure) @bsoyka can help you
> > 
> > 
> > I enforce Fatih's suggestion
> 
> @esau-morais what do you mean by that?

I'm just agreeing that it'd would be great to apply your suggestion

File: ci/check-registry.js (hunk 2 above)
Comment (haliliceylan) [reply ci/check-registry.js]: fixed

File: ci/check-registry.js (hunk 1 above)
Comment (haliliceylan) [reply ci/check-registry.js]: fixed

**Comment by haliliceylan** [comment]
> @haliliceylan thank you for the contribution. Do you mind putting this check into a function and adding a test for it? If you need help, I and (I am pretty sure) @bsoyka can help you

I can move it to a separate function. I would love to write to tests !

Test scenarios can be as follows;
- correct scenario, 2 different github usernames, 2 different domains
- wrong scenario, 2 same github username, 2 different domains

Is there a test written for CI Script as an example in this repository?

**Comment by fatih-yavuz** [comment]
> > > > @haliliceylan thank you for the contribution. Do you mind putting this check into a function and adding a test for it? If you need help, I and (I am pretty sure) @bsoyka can help you
> > > 
> > > 
> > > I enforce Fatih's suggestion
> > 
> > 
> > @esau-morais what do you mean by that?
> 
> I'm just agreeing that it'd would be great to apply your suggestion

Thank you. I wasn't sure whether you were applying the suggestion or just supporting the idea. Thanks for the explanation

**Comment by fatih-yavuz** [comment]
> > @haliliceylan thank you for the contribution. Do you mind putting this check into a function and adding a test for it? If you need help, I and (I am pretty sure) @bsoyka can help you
> 
> I can move it to a separate function. I would love to write to tests !
> 
> Test scenarios can be as follows;
> 
> * correct scenario, 2 different github usernames, 2 different domains
> * wrong scenario, 2 same github username, 2 different domains
> 
> Is there a test written for CI Script as an example in this repository?

No, but I will set it up in a few hours

**Comment by fatih-yavuz** [comment]
I see that you already did 🚀

**Comment by haliliceylan** [comment]
I separated my function, also  I created  a separate file for my function. Because If I try to import the "check-registry" file, main() function is working and gives an error. So I created a new folder for registry file related checks then I put my script into it.

I installed the jest package.

I create a GitHub workflow file for when "ci" folder changes, it will run all tests under "ci" folder. idk is working or not; In my local I tried with "act" command (local github workflow simulator) and it seems ok

**Comment by haliliceylan** [comment]
oh i forget to add `pull_request` in workflow file...

**Comment by fatih-yavuz** [comment]
> oh i forget to add `pull_request` in workflow file...

I fixed it now. I am merging it. Thank you

**Pull request merged**